Fix to original PR - Now sends FIRMWARE_RESTART before powering on printer, queries "webhooks" printer object, sends model when state == 'ready'

truckershitch opened the pull request:
Before this fix, if the user requested the model to start printing immediately, the plugin might upload the model before the printer was ready to accept the print job.  This change should fix that issue -- it now queries the "webhooks" printer object and waits until state == 'ready' before sending the model.

At the moment, the queries occur every 0.5 seconds for a total of 10 seconds before a timeout occurs.

Thanks again for a great plugin!

emtrax-ltd commented:
attention: it was/is a feature of the plugin to upload the code while the printer itself is offline. with the fix for the pr it looks like you are enforced to turn on the printer before an upload is started.

truckershitch commented:
What seems to happen here with my printer and setup is this:

When a print job finishes, a macro turns the power to the smart plug off.  Then the printer is off.  In my particular case, my printer (an Ender-3 v2 if that matters) is connected by USB to my Raspbery Pi 3B+.

At that point, the serial device disconnects and that device is gone until the printer power is turned back on.  Klipper/Klippy cannot reach the printer and quits with an error, saying it cannot reach the mcu.

If the printer is turned back on, Klipper/Klippy still refuses to accept uploads (at least the way your plugin sends them) and the connection times out.  If a FIRMWARE_RESTART is sent, Klipper/Klippy will again try and connect to the printer.  So that is what I did in this "update" to the PR.

There may be, and there probably is, another solution.

What if there was a decision made in your plugin by the configuration options chosen?  As in, "Is your printer connected with a smart plug?   If so, what is the device name in moonraker.conf?"  Then, you plugin could choose how the code was uploaded.

But in answer to your comment:  As I see it here locally with my printer, if the printer is not already powered on (and I _think_ it has to be state == 'ready'), the code will not upload at all and there will be a timeout.

I am not trying to take over your work!

Do you have any suggestions?    I am very impressed with what you have done so far.

truckershitch commented:
OH!  Okay -- so the way I wrote it, it will do a FIRMWARE_RESTART even if the printer is running.

You know, I was so glad that I fixed that problem today that I created another one!

So I propose to only do a FIRMWARE_RESTART if a timeout occurs.  As I see it, if a timeout is not a network problem, the printer is probably powered off?  Does that make more sense?

Since it is a powerful command, maybe there should be a confirmation from the user (click this button to do a FIRMWARE_RESTART).

I think both of those are good.  I'll take a look at it tomorrow morning.

How does that sound to you?

emtrax-ltd commented:
i think we can imply your suggested behavior of the upload process with the checkbox-state of "start print job" - if enabled, then (if needed) startup printer, reset firmware, upload and print will start. if not enabled, then just upload to the box/pi (there is no need to startup anything).

emtrax-ltd commented:
there is a smart trick for an automatically "restart-firmwere" if the pi detects the board (mcu) and will be available at the usb-port.

/etc/udev/rules.d/98-klipper.rules

```
## rule to restart klipper when the printer is connected via usb
SUBSYSTEM=="usb", ATTRS{idVendor}=="1d50", ATTRS{idProduct}=="614e", ACTION=="add", RUN+="/bin/su pi -c '/bin/echo FIRMWARE_RESTART > /tmp/printer'"

```
-> https://github.com/KevinOConnor/klipper/issues/835
-> you have to configure your correct vendor- and product-id -> get it via "lsusb"

truckershitch commented:
I changed the logic flow and it is working pretty well so far.

When rebooting the Raspberry Pi and the printer power is OFF, querying 'printer/info' returns:

```
state_message == 'Can not update MCU \'mcu\' config as it is shutdown\nOnce the underlying issue is corrected, use the\n"FIRMWARE_RESTART" command to reset the firmware, reload the\nconfig, and restart the host software.\nError configuring printer\n'

state == 'error'
```

If the code sees state == 'error', it (should) make a new dialog status message with (firmware_restart/cancel) buttons.  I have not been able to test this "case" in the last few trials -- the code checks to see the power status, and if off, starts the printer (no error is returned)  and if on, merely starts your file upload routine as usual.

At this point, I do not know if the new dialog status message will work with two buttons or not as I haven't had that 'error' status since I wrote it.  Moonraker had an update today and I don't know if that is related or not.

I will do more testing here.

truckershitch commented:
Hey @emtrax-ltd :)

I cleaned my code up.  I tried to follow your existing conventions (with the underscore-prefix "_private" methods and their placement in your code) and I also tried to place my additional modules in good places vertically in the source code files.

I got rid of the FIRMWARE_RESTART calls.  I also added the suggested udev rule.  That seems to be working well.  Thank you for that suggestion.

Sometimes, Klipper shuts down.  This seems to happen when it cannot reach the printer at all.  At that point, Moonraker does not want to answer calls to its API  (as far as I can tell) so user intervention is necessary.  I left that up to the user.

I think the code flows better now.  Let me know what you think.

emtrax-ltd commented:
OK... code looks clean and functional - i will test it next 2 days. I want to make the following suggestion:

    def onFilenameAccepted(self):
    ...
        if self._power_device and self._startPrint and self._startPrint == True:
            self.getPrinterDeviceStatus()
        else:
            self.getPrinterInfo()
    ...

I think it's not necessary to turn on the printer if you don't want to start a print-job in this place. What do you think about it?

truckershitch commented:
Yes, I agree.  I made a few changes there.

In my local testing, I find that when the printer power is OFF for a few minutes, Klipper will not respond to the 'printer/info' GET request.  In that case, if a FIRMWARE_RESTART is sent, Klipper will then respond and continue, and the job can be sent to Moonraker without turning the printer ON.

The code checks first to see if the printer status is OFF.  If it is and self._start_print is True, the printer power is set to ON.  Otherwise the printer power remains OFF.

I believe:
```
if self._startPrint and self._startPrint == True:
```
is equivalent to:
```
if self._startPrint:
```

as None evalutates to False.  But that is a very small thing.

https://switowski.com/blog/checking-for-true-or-false has a bit to say on the speed of evaluating boolean expressions (just found this this morning).

I started a print job this morning when the printer was OFF and after I fixed a few typos and bugs, it was sent to Moonraker and the printer did not turn ON.  The changes seem to be working.

See what you think!

c-ridgway commented:
I've noticed /tmp/printer doesn't exist anymore. Doing this worked instead:
`echo RESTART > /home/*/printer_data/comms/klippy.serial`
